### PR TITLE
Refactor parseFilter to only flatten filter entries that need it

### LINKED ISSUE #9746: Moving or zooming in map layout triggers an error
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

Moving or zooming seems to trigger this error each time:

https://user-images.githubusercontent.com/42867097/141438339-9fe2ffd4-f057-4baa-ab90-32daba9323bd.mp4

Might be related to the fix in #9615.

### To Reproduce

Should be reproducible on map layout without any data (at least in my case):

https://user-images.githubusercontent.com/42867097/141438168-cb211e8d-fe47-4ab9-994b-03c3376b5ea0.mp4

### Errors Shown

Whenever the error popup occurs, a GET request gets a 400 response.

Request URL:

```js
https://<url here>/items/locations?limit=1000&fields[]=location&fields[]=name&fields[]=id&sort[]=id&page=1&filter=%7B%22location%22:%7B%22_intersects_bbox%22:%7B%22type%22:%22Polygon%22,%22coordinates%22:[[[-91.87273857268036,44.250154169852266],[-66.82128551912203,44.250154169852266],[-66.82128551912203,61.390294098284215],[-91.87273857268036,61.390294098284215],[-91.87273857268036,44.250154169852266]]]%7D%7D%7D&meta[]=filter_count&meta[]=total_count
```

Formatted Query Strings from above:

```
limit: 1000
fields[]: location
fields[]: name
fields[]: id
sort[]: id
page: 1
filter: {
   "location":{
      "_intersects_bbox":{
         "type":"Polygon",
         "coordinates":[
            [
               [
                  -91.87273857268036,
                  44.250154169852266
               ],
               [
                  -66.82128551912203,
                  44.250154169852266
               ],
               [
                  -66.82128551912203,
                  61.390294098284215
               ],
               [
                  -91.87273857268036,
                  61.390294098284215
               ],
               [
                  -91.87273857268036,
                  44.250154169852266
               ]
            ]
         ]
      }
   }
}
meta[]: filter_count
meta[]: total_count
```

Response Body:

```js
errors: [
  {
    "message": "\"_intersects_bbox\" has to be a valid GeoJSON object",
    "extensions": {
      "code": "INVALID_QUERY"
    }
  }
]
```

### What version of Directus are you using?

9.0.1

### What version of Node.js are you using?

16.13.0

### What database are you using?

PostgreSQL 13.4

### What browser are you using?

Chrome

### What operating system are you using?

Windows

### How are you deploying Directus?

Docker

### PR DESCRIPTION
**Pull request by oreilles**
Alternative to https://github.com/directus/directus/pull/9753
Fixes https://github.com/directus/directus/issues/9746

~~@rijkvanzanten I replaced the `get` function you wrote with the one from `lodash`. The difference is that if you get `$USER.data.foo` and `$USER.data`is an array, yours returns `$USER.data.map(entry => entry['foo'])` wherehas `lodash`'s will return `$USER.data['foo']`.  Is that scenario actually relevant ?~~
Nevermind, since it probably was necessary I left it in place, just refactored it too.

### CODE REVIEW

File: packages/shared/src/utils/parse-filter.ts [hunk 1]
```diff
@@ -2,80 +2,66 @@ import { REGEX_BETWEEN_PARENS } from '../constants';
...
+		if (value === '$CURRENT_ROLE') return accountability?.role ?? null;
+		return get(context, value, null);
 	}
 }
+
+function get(object: Record<string, any> | any[], path: string, defaultValue: any): any {
+	const [key, ...rest] = path.split('.');
+	const result = Array.isArray(object) ? object.map((entry) => entry[key!]) : object?.[key!];
+	return rest.length === 0 ? result ?? defaultValue : get(result, rest.join('.'), defaultValue);
+}
```
Comment (rijkvanzanten) [inline packages/shared/src/utils/parse-filter.ts:67]: While this is a very smart implementation, it's way harder to read and follow along what's going on. Is there any practical benefit to this refactor? I'd prefer keeping the older implementation for maintenance reasons if it's only a stylistic change.

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]

**Comment by rijkvanzanten** [comment]
> Is that scenario actually relevant ?

Yes, very 😄 That's the reason to have a custom get function instead of relying in lodash 👍🏻

File: packages/shared/src/utils/parse-filter.ts [hunk 2]
```diff
@@ -2,80 +2,69 @@ import { REGEX_BETWEEN_PARENS } from '../constants';
...
-		} else if (value && typeof value === 'object') {
-			return value[key];
-		}
+function get(object: Record<string, any> | any[], path: string, defaultValue: any): any {
+	const [key, ...follow] = path.split('.');
+	const result = Array.isArray(object) ? object.map((entry) => entry[key!]) : object?.[key!];
+	if (follow.length > 0) {
+		return get(result, follow.join('.'), defaultValue);
 	}
+	return result;
```
Comment (Oreilles) [inline packages/shared/src/utils/parse-filter.ts:69]: @rijkvanzanten you're right, It's probably a bit better like this. 🙂 I feel like reducing the locs and avoiding nested functions makes this one easier to maintain than the previous one.

Comment (rijkvanzanten) [reply packages/shared/src/utils/parse-filter.ts:69]: Reducing LOCs is not always equal to making it easier to read / understand 🙂 Agreed on the nested functions though, that's not always as easy to follow 👍🏻

**Comment by Oreilles** [comment]
E2E tests failed for MSSQL due to an issue unrelated to the code:
https://github.com/directus/directus/runs/4212923354?check_suite_focus=true
```
[14:50:33] MS SQL Server [started]

/home/runner/work/directus/directus/node_modules/tedious/lib/connection.js:1381
      this.emit('error', new _errors.ConnectionError(message, 'ESOCKET'));
                         ^
ConnectionError: Connection lost - unexpected end of message stream
    at Connection.socketError (/home/runner/work/directus/directus/node_modules/tedious/lib/connection.js:1381:26)
    at /home/runner/work/directus/directus/node_modules/tedious/lib/connection.js:2378:14
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
Emitted 'error' event on Connection instance at:
    at Connection.emit (/home/runner/work/directus/directus/node_modules/tedious/lib/connection.js:1031:18)
    at Connection.socketError (/home/runner/work/directus/directus/node_modules/tedious/lib/connection.js:1381:12)
    at /home/runner/work/directus/directus/node_modules/tedious/lib/connection.js:2378:14
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  code: 'ESOCKET'
}
Error: Process completed with exit code 1.
```

**Review by licitdev — COMMENTED** [review]
Closed #9753 in favour of this PR.

@Oreilles Really impressive refactor! 👍🏻 
The `flatMap` works well in flattening the array from dynamic variable.

File: packages/shared/src/utils/parse-filter.ts (hunk 1 above)
Comment (Oreilles) [reply packages/shared/src/utils/parse-filter.ts]: Is it fine with the change I made or do you want to revert to the previous implementation ?

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**